### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by atarix83**
This is a new PR that should resolve build issue occurred in the previous one https://github.com/DSpace/dspace-angular/pull/230

### CODE REVIEW

**Comment by atarix83** [comment]
@artlowel @tdonohue 

Now with this new PR travis-ci build works properly.

It's ready to be merged if there aren't other comments.

File: src/app/core/auth/server-auth.service.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  public authenticate(user: string, password: string): Observable<AuthStatus> {
+    // Attempt authenticating the user using the supplied credentials.
+    const body = encodeURI(`password=${password}&user=${user}`);
+    const options: HttpOptions = Object.create({});
+    let headers = new HttpHeaders();
+
+    // NB this could be use to avoid the problem with the authentication is case the UI is rendered by Angular Universal.
+    const clientIp = this.req.connection.remoteAddress;
+
+    headers = headers.append('Content-Type', 'application/x-www-form-urlencoded');
```
Comment (tomdesair) [inline src/app/core/auth/server-auth.service.ts]: Could you also add 
```
headers = headers.append('X-Forwarded-For', clientIp);
``` 
here? That can already help solving the "disable ip check" problem in the REST backend.

Comment (atarix83) [reply src/app/core/auth/server-auth.service.ts]: @tomdesair 

done

File: src/app/core/auth/server-auth.service.ts [hunk 2]
```diff
@@ -16,32 +17,29 @@ import { CheckAuthenticationTokenAction } from './auth.actions';
...
+    // Determine if the user has an existing auth session on the server
     const options: HttpOptions = Object.create({});
     let headers = new HttpHeaders();
 
-    // NB this could be use to avoid the problem with the authentication is case the UI is rendered by Angular Universal.
-    const clientIp = this.req.connection.remoteAddress;
+    headers = headers.append('Accept', 'application/json');
+    headers = headers.append('Authorization', `Bearer ${token.accessToken}`);
+    // NB this is used to pass server client IP check.
+    const clientIp = this.req.get('x-forwarded-for');
```
Comment (tomdesair) [inline src/app/core/auth/server-auth.service.ts]: For requests that do not provide a `X-Forwarded-For` header, I would fallback to `this.req.connection.remoteAddress`. So something like:

```
if(isBlank(clientIp)) {
   clientIp = this.req.connection.remoteAddress;
}
```

**Comment by tomdesair** [comment]
1. I've added a request in the code for one more small change.

2. I tested this pull-request locally. I can successfully authenticate using the login-popup. After filling in my credentials, my name appears in the top menu bar. **However**, if I then refresh the page with the browser it seems that I'm logged out again: the top menu bar displays "Log In" instead of my name. Is this a bug or did I configure something wrong?

**Comment by atarix83** [comment]
> I tested this pull-request locally. I can successfully authenticate using the login-popup. After filling in my credentials, my name appears in the top menu bar. However, if I then refresh the page with the browser it seems that I'm logged out again: the top menu bar displays "Log In" instead of my name. Is this a bug or did I configure something wrong?

This seems the consequence of the ip check issue, please could you check if it is enabled on server REST that you're using?

**Comment by tomdesair** [comment]
I've disabled the server side ip check and I still have the same issue. I created a short screen recording here: https://dspace-org.slack.com/files/U3SK685FA/FBJB7Q6BT/dspace7-angular-login-refresh-issue.mp4

**Comment by tomdesair** [comment]
Hi @atarix83,

That indeed solved the problem. I can now safely refresh the page after logging in. I can even leave the _server-side ip-check enabled_ now! Good job!

So from a functional perspective, this looks fine to me.

**Pull request merged**